cpp/src/io/text/multibyte_split.cu

```diff
@@ -275,180 +517,126 @@ int64_t multibyte_split_scan_full_source(cudf::io::text::data_chunk_source const
     tile_multistates,
     tile_offsets,
     multistate_seed,
-    0);
+    {});
 
-  fork_stream(streams, stream);
+  auto reader               = source.create_reader();
+  auto chunk_offset         = std::max<int64_t>(0, byte_range.offset() - delimiter.size());
+  auto const byte_range_end = byte_range.offset() + byte_range.size();
+  reader->skip_bytes(chunk_offset);
+  output_builder<int64_t> offset_storage(ITEMS_PER_CHUNK / delimiter.size() + 1, stream);
+  output_builder<char> char_storage(ITEMS_PER_CHUNK, stream);
 
-  auto reader = source.create_reader();
+  fork_stream(streams, stream);
 
   cudaEvent_t last_launch_event;
   cudaEventCreate(&last_launch_event);
 
-  for (int32_t i = 0; true; i++) {
-    auto base_tile_idx = i * TILES_PER_CHUNK;
-    auto chunk_stream  = streams[i % streams.size()];
-    auto chunk         = reader->get_next_chunk(ITEMS_PER_CHUNK, chunk_stream);
+  auto& read_stream     = streams[0];
+  auto& scan_stream     = streams[1];
+  auto chunk            = reader->get_next_chunk(ITEMS_PER_CHUNK, read_stream);
+  int64_t base_tile_idx = 0;
+  std::optional<int64_t> first_offset;
+  std::optional<int64_t> last_offset;
+  if (byte_range.offset() == 0) { first_offset = 0; }
+  std::swap(read_stream, scan_stream);
 
-    if (chunk->size() == 0) { break; }
+  while (chunk->size() > 0) {
+    // if we found the last delimiter, or didn't find delimiters inside the byte range at all: abort
+    if (last_offset || (!first_offset && chunk_offset >= byte_range_end)) { break; }
```

Comment:
Just because both `int` and `optional` implicitly cast to bool :(
```suggestion
    if (last_offset.has_value() || (!first_offset && chunk_offset >= byte_range_end)) { break; }
```